### PR TITLE
Added diff option to compare processed profiles.

### PR DESCRIPTION
**Pull request by nandor**
`prof_spacetime diff profile_1 profile2` compares the number of memory blocks live at program exit time in each file in an application, producing an output along the lines of:

```
File                                                    Words    Allocations         Blocks
src/utils.ml                                          -753885        -765543        -225188
src/stdune/path.ml                                    -323112      -13199839        -107428
src/stdune/map.ml                                       +2496           -762           +836
string.ml                                             +139383        +325271         +21483
pervasives.ml                                         +449678       +1514358         -40923
```

### CODE REVIEW

**Comment by lpw25** [comment]
Sorry for the slow response. I've found time to review this now. 

I'm a little unsure about defaulting to splitting things by file. None of the other commands work in terms of files. I assume that you did this because you wanted to compare two different implementations of a file, and so the individual allocations did not have the same location in the two profiles. I can't think of a simple better approach, so I guess its fine for now.

File: bin/prof_spacetime.ml [hunk 1]
```diff
@@ -11,6 +11,9 @@ type command =
         print_symbol:      bool;
         print_line_number: bool; }
   | Process
+  | Diff of
```
Comment (lpw25) [inline bin/prof_spacetime.ml:14]: This command should be able to run on either processed or unprocessed profiles, based on the `-p` option, just like the other commands. So I would expect to see a `processed` field here like in the other cases.

File: src/diff.ml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+module AllocMap = Map.Make(String)
+module NameSet = Set.Make(String)
+
+
+let get_alloc_map snapshot =
+  let rec loop snapshot acc =
+    Section.fold Section.(fun addr item acc ->
+      let value = Item.value item in
+
```
Comment (lpw25) [inline src/diff.ml]: Please no blank lines within expressions

File: src/diff.ml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      let words = Allocation.words value in
+      let allocations = Allocation.allocations value in
+      let blocks = Allocation.blocks value in
+
+      try
+        let display = Item.display item in
+        let name = String.sub display 0 (String.index display ':') in
+        acc |> AllocMap.update name (fun v -> Some (match v with
```
Comment (lpw25) [inline src/diff.ml]: I'd prefer to avoid using `|>` in general and especially for just passing an argument like this.

File: src/diff.ml [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+          | Some v ->
+            { words = words + v.words
+            ; allocations = allocations + v.allocations
+            ; blocks = blocks + v.blocks
+            }))
+      with
+        Not_found ->
+          loop (Item.select item) acc
+    ) snapshot acc
+  in loop snapshot AllocMap.empty
```
Comment (lpw25) [inline src/diff.ml]: House style: `in` on its own line unless it is on the same line as `let`.

File: src/diff.ml [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+let get_keys map =
+  AllocMap.fold (fun k _ acc -> NameSet.add k acc) map NameSet.empty
+
+let diff_maps map0 map1 =
+  let keys = NameSet.union (get_keys map0) (get_keys map1) in
+  let diffs = NameSet.elements keys
```
Comment (lpw25) [inline src/diff.ml]: There needs to be a newline after `diffs =` otherwise it is awkward to read.

**Comment by nandor** [comment]
I defaulted to file-level granularity as the line numbers of allocation points are bound to change between two different versions and the comparisons would be quite useless. File names closely match modules and I found them to be quite useful.

**Review by lpw25 — CHANGES_REQUESTED** [review]
A bit of style nitpicking, and the `-p` option should be respected. Once that's addressed its good to go.

**Comment by lpw25** [comment]
> I defaulted to file-level granularity as the line numbers of allocation points are bound to change between two different versions and the comparisons would be quite useless. File names closely match modules and I found them to be quite useful.

Yeah, that's what I figured. I think function-level granularity would fit better with the rest of the tool, and it should be possible, but it is much harder to implement so I think file-level is fine for now.

**Comment by nandor** [comment]
Do the spacetime profiles capture function information though? I haven't found any methods to retrieve function names. 

I have addressed your style comments.

**Comment by lpw25** [comment]
> Do the spacetime profiles capture function information though?

Yes and no. IIRC we have the address of the call site, and the dwarf info can give you the symbol of the function containing that address. So we could get it if we wanted to as long as the user used the `-e` option. It would be a reasonable amount of work though. Not worth it for this simple PR -- just thinking out loud about how we might do things in an ideal world.

**Comment by lpw25** [comment]
>  It would be a reasonable amount of work though.

Actually, I'd forgotten that this is already implemented in spacetime_lib. It's not exposed well in the model in this repo -- it only appears as part of the location string -- but it wouldn't be hard to expose it properly.

**Comment by lpw25** [comment]
All looks good.

**Pull request merged**